### PR TITLE
feat(weight_sync): add unified GCSWeightSync and GCSWeightSyncHandler…

### PR DESCRIPTION
**Pull request by s-noghabi**
… with checksum verification

Resolves #\<issue_number_goes_here\>

> It's a good idea to open an issue first for discussion.

**Reference**

**Colab Notebook**

**Checklist**

- [ ] I have added all the necessary unit tests for my change.
- [ ] I have verified that my change does not break existing code and all unit tests pass.
- [ ] I have added all appropriate doc-strings/documentation.
- [ ] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [ ] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [ ] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

> **Note**: Standard CPU unit tests, package builds, and documentation checks will run automatically on pull requests. Once the PR is approved and ready for submission, maintainers will add the `ready-to-submit` label to trigger full TPU testing.

### CODE REVIEW

File: tunix/experimental/examples/math_gsm8k_dist/launcher.sh [hunk 1]
```diff
@@ -652,6 +655,9 @@ echo "Launching rollout node with sampler=$SAMPLER on TPU chips $ROLLOUT_TPU_CHI
   export TPU_HOST_BOUNDS=${TPU_HOST_BOUNDS}
   export LIBTPU_INIT_ARGS="--deepsea_chips_per_host_bounds=${TPU_CHIPS_PER_HOST_BOUNDS} --deepsea_host_bounds=${TPU_HOST_BOUNDS}"
   export VERIFY_WEIGHTS=${VERIFY_WEIGHTS}
+  export WEIGHT_SYNC_MODE=${WEIGHT_SYNC_MODE}
+  export WEIGHT_SYNC_GCS_DIR=${WEIGHT_SYNC_GCS_DIR}
+  export VLLM_ALLOW_INSECURE_SERIALIZATION=1
```
Comment (s-noghabi) [inline tunix/experimental/examples/math_gsm8k_dist/launcher.sh]: why did we add this?

File: tunix/experimental/rollout/vllm_sampler_adapter.py [hunk 2]
```diff
@@ -427,31 +436,81 @@ async def pre_weight_sync(
...
+        logger.info(
+            "Executing GCS weight_sync restore from %s on TPU workers...",
+            ckpt_path,
+        )
+        checksums = await sampler.gcs_h2d(
+            checkpoint_path=ckpt_path,
+            source_checksums=src_checksums,
+        )
+      else:
+        logger.info("Executing weight_sync barrier on Raiden synchronizers...")
```
Comment (s-noghabi) [inline tunix/experimental/rollout/vllm_sampler_adapter.py:501]: why we added this?

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 3]
```diff
@@ -815,6 +819,10 @@ def _claim(key: str):
     runner.state = jax.tree_util.tree_unflatten(
         jax.tree_util.tree_structure(runner.state), new_leaves
     )
+    if hasattr(runner, "refresh_state_leaves") and callable(
+        runner.refresh_state_leaves
+    ):
+      runner.refresh_state_leaves()
```
Comment (s-noghabi) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: why add this?

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 4]
```diff
@@ -835,17 +843,34 @@ def work_unit_metadata_all(self) -> List[weight_sync.WorkUnitMetadata]:
...
+    limit = len(self.arrays) if verify_all else min(sample, len(self.arrays))
+
+    head: dict[str, Any] = {}
+    for idx in range(limit):
+      name = self.names[idx]
+      val = float(per_tensor[idx])
+      key = _param_key(name) if verify_all else name
+      head[key or name] = val
+
+    head["__grand_total__"] = float(per_tensor.sum())
```
Comment (s-noghabi) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: why change these?

File: tunix/experimental/weight_sync/raiden_synchronizer.py [hunk 5]
```diff
@@ -938,3 +976,11 @@ def _patched_apply_to_runner(self, runner: Any) -> None:
     )
   except AttributeError as e:
     logging.debug("tpu_inference not available to patch: %s", e)
+
+  try:
```
Comment (s-noghabi) [inline tunix/experimental/weight_sync/raiden_synchronizer.py]: why add these?

File: tunix/experimental/weight_sync/weight_sync.py [hunk 6]
```diff
@@ -316,6 +336,8 @@ class WeightSyncHandler(abc.ABC):
   units are registered and transfers run.
   """
 
+  needs_manifest_preflight: bool = True
```
Comment (s-noghabi) [inline tunix/experimental/weight_sync/weight_sync.py]: what's this for?

File: tunix/experimental/weight_sync/weight_sync.py [hunk 7]
```diff
@@ -521,3 +550,212 @@ async def get_weight_sync_status(self) -> Mapping[str, Any]:
...
+
+  def release(self, sync_request: Any = None) -> None:
+    """Releases or cleans up per-round transport staging resources."""
+    del sync_request
+
+  def close(self) -> None:
+    """Closes the synchronizer and releases persistent resources."""
+
+
+def verify_weight_checksums(
```
Comment (s-noghabi) [inline tunix/experimental/weight_sync/weight_sync.py:627]: is this guarded by a flag?

**Review by niting — APPROVED** [review]
